### PR TITLE
Sync file_read, file_write from Nuttx

### PR DESCRIPTION
**Pull request by junmin-kim**

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [be2e5a69cbd231ad2517a640423df748013929b9] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [be2e5a69cbd231ad2517a640423df748013929b9] - Code Rule Check Result:
os/fs/vfs/fs_write.c:218: ERROR: [SPC_M_OPR] spaces required around that ':' (ctx:VxE)

File: os/fs/vfs/fs_pread.c [hunk 1]
```diff
@@ -112,7 +112,10 @@ ssize_t file_pread(FAR struct file *filep, FAR void *buf, size_t nbytes, off_t o
 	/* Then perform the read operation */
 
 	ret = file_read(filep, buf, nbytes);
-	errcode = get_errno();
+	if (ret < 0) {
```
Comment (Taejun-Kwon) [inline os/fs/vfs/fs_pread.c]: I think set_errno is already being called in file_read & file_seek. how about below?
```
if (ret < 0) {
    ret = ERROR;
}
...
//set_errno(errcode);  //remove this line!

File: os/fs/vfs/fs_read.c [hunk 2]
```diff
@@ -111,23 +117,16 @@ ssize_t file_read(FAR struct file *filep, FAR void *buf, size_t nbytes)
 	 * method?
 	 */
 
-	else if (inode && inode->u.i_ops && inode->u.i_ops->read) {
+	else if (inode != NULL && inode->u.i_ops && inode->u.i_ops->read) {
```
Comment (Taejun-Kwon) [inline os/fs/vfs/fs_read.c]: same as previous.

File: os/fs/vfs/fs_read.c [hunk 3]
```diff
@@ -111,23 +117,16 @@ ssize_t file_read(FAR struct file *filep, FAR void *buf, size_t nbytes)
...
 		 * mountpoint, we depend on the read methods being identical in
 		 * signature and position in the operations vtable.
 		 */
 
-		ret = (int)inode->u.i_ops->read(filep, (char *)buf, (size_t)nbytes);
-	}
-
-	/* If an error occurred, set errno and return -1 (ERROR) */
-
-	if (ret < 0) {
```
Comment (Taejun-Kwon) [inline os/fs/vfs/fs_read.c]: uh oh, you removed code here, then file_seek and another function also need to be checked.

**Comment by seinfra** [comment]
Target : [22d151265004b5550e98997bd5f6bb142766e608] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [22d151265004b5550e98997bd5f6bb142766e608] - Code Rule Check OK.

File: os/fs/vfs/fs_read.c [hunk 4]
```diff
@@ -111,23 +117,16 @@ ssize_t file_read(FAR struct file *filep, FAR void *buf, size_t nbytes)
...
-		set_errno(-ret);
```
Comment (chanijjani) [inline os/fs/vfs/fs_read.c]: As Taejun said, it seems tgat if this set_errno() is removed, we need to add the set_errno() to every place on which file_read is called..
Are there some reasons that this removal is better?

File: os/fs/vfs/fs_write.c [hunk 5]
```diff
@@ -80,45 +80,47 @@
...
-	ret = inode->u.i_ops->write(filep, buf, nbytes);
-	if (ret < 0) {
-		err = -ret;
-		goto errout;
-	}
-
-	return ret;
-
-errout:
-	set_errno(err);
```
Comment (chanijjani) [inline os/fs/vfs/fs_write.c]: Similar to the comment in file_read().

File: os/include/tinyara/fs/fs.h [hunk 6]
```diff
@@ -826,35 +826,58 @@ int fs_getfilep(int fd, FAR struct file **filep);
...
+ *   buf    - User-provided to save the data
+ *   nbytes - The maximum size of the user-provided buffer
+ *
+ * Returned Value:
+ *   The positive non-zero number of bytes read on success, 0 on if an
+ *   end-of-file condition, or a negated errno value on any failure.
  *
  ****************************************************************************/
 
-#if CONFIG_NFILE_DESCRIPTORS > 0
```
[the remark this answers is not in the document]
Comment (junmin-kim) [reply os/include/tinyara/fs/fs.h]: In `config.h`,
```c++
#ifndef CONFIG_NFILE_DESCRIPTORS
# define CONFIG_NFILE_DESCRIPTORS 0
#endif
```
So, CONFIG_NFILE_DESCRIPTORS always >= 0. So the statement is not needed.

File: os/include/tinyara/fs/fs.h [hunk 7]
```diff
@@ -826,35 +826,58 @@ int fs_getfilep(int fd, FAR struct file **filep);
...
+ *
+ * Returned Value:
+ *  On success, the number of bytes written are returned (zero indicates
+ *  nothing was written).  On any failure, a negated errno value is returned
+ *  (see comments withwrite() for a description of the appropriate errno
+ *  values).
  *
  ****************************************************************************/
 
-#if CONFIG_NFILE_DESCRIPTORS > 0
```
[the remark this answers is not in the document]
Comment (junmin-kim) [reply os/include/tinyara/fs/fs.h]: In `config.h`,
```c++
#ifndef CONFIG_NFILE_DESCRIPTORS
# define CONFIG_NFILE_DESCRIPTORS 0
#endif
```
So, CONFIG_NFILE_DESCRIPTORS always >= 0. So the statement is not needed.

File: os/fs/vfs/fs_read.c (hunk 4 above)
Comment (junmin-kim) [reply os/fs/vfs/fs_read.c]: 1. I sync the Nuttx code. For future porting, following the Nuttx code is a good choice.
2. The errno should be set just before the function returns. errno can be polluted by other functions.

File: os/fs/vfs/fs_write.c (hunk 5 above)
Comment (junmin-kim) [reply os/fs/vfs/fs_write.c]: Same above.

**Comment by seinfra** [comment]
Target : [eeb37fe40330ee55b99bc769fa58ed9b012ba90d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [eeb37fe40330ee55b99bc769fa58ed9b012ba90d] - Code Rule Check OK.

File: os/fs/vfs/fs_pwrite.c [hunk 8]
```diff
@@ -111,7 +111,10 @@ ssize_t file_pwrite(FAR struct file *filep, FAR const void *buf, size_t nbytes,
 	/* Then perform the write operation */
 
 	ret = file_write(filep, buf, nbytes);
-	errcode = get_errno();
+	if (ret < 0) {
```
Comment (zhouxinhe) [inline os/fs/vfs/fs_pwrite.c]: Please also take care below code snippet in aio_wrint.c at line 182 ~ 189.
```c
	/* Check the result of the write */

	if (nwritten < 0) {
		int errcode = get_errno();
		fdbg("ERROR: write/pwrite failed: %d\n", errcode);
		DEBUGASSERT(errcode > 0);
		aiocbp->aio_result = -errcode;
	} ...
```

Comment (zhouxinhe) [reply os/fs/vfs/fs_pwrite.c]: In addition, on nuttx, file_pwrite() does not set errno, but return negated error code as file_write().

**Pull request merged**